### PR TITLE
DAOS-5528 placement: use array for grp_map tracking

### PR DESCRIPTION
**Pull request by wangdi1**
Grp_map is allocated by object layout number, but setbit
by target number, which might overwrite the grp_map.

let's Using array, instead of bitmap, for grp_map tracking,
because bitmap + target id might need allocate bitmaps
for all targets in pool map.

Signed-off-by: Di Wang <di.wang@intel.com>

### CODE REVIEW

File: src/placement/pl_map_common.c [hunk 1]
```diff
@@ -327,59 +327,103 @@ determine_valid_spares(struct pool_target *spare_tgt, struct daos_obj_md *md,
...
+			D_FREE_PTR(f_shard);
+			continue;
+		}
 
-		current = current->next;
-	}
+		if (grp_map_idx > grp_map_size) {
+			uint32_t *new_grp_map;
 
+			new_grp_map = grp_map_extend(grp_map, &grp_map_size);
```
Comment (liuxuezhao) [inline src/placement/pl_map_common.c:411]: originally the grp_map point to stack memory, how can it be extend (realloc) later in grp_map_extend?

Comment (liuxuezhao) [reply src/placement/pl_map_common.c:411]: Ok, i see.

**Review by liuxuezhao — APPROVED** [review]

**Comment by wangdi1** [comment]
> Looks good to me. My only concern is that this takes this from an O(n) algorithm to O(n^2), because it's now having to search the list each time rather than just doing an array index to get whether the bitmap value is set or not. I think that's probably fine for this usage (only called during addition/extension)

right, it might takes  a bit more time.  But keeps using bitmap will need allocate bits for all targets inside the pool, which might be huge, i.e. allocate a lot space for each object extension. 

Given that the layout number for each object is relatively small, so I just use array, probably fine for this case, since this is not in the hot path anyway.

**Comment by johannlombardi** [comment]
Have you guys looked into test failures? Are they known?

**Comment by wangdi1** [comment]
> Have you guys looked into test failures? Are they known?

yes, they are known

**Pull request merged**